python/prairielearn.py

```diff
@@ -167,14 +168,19 @@ def from_json(v):
     return v
 
 
-def inner_html(element):
-    inner = element.text
-    if inner is None:
-        inner = ''
-    inner = html.escape(str(inner))
-    for child in element:
-        inner += lxml.html.tostring(child, method='html').decode('utf-8')
-    return inner
+def inner_html(element: lxml.html.HtmlElement) -> str:
+    """
+    Gets the inner HTML of an element. A bit ugly, but hacked together to be as fast as possible
+    (since this is a function that gets used in a ton of places). The performance increase from
+    the old version of this function could have a substantial impact on question render times.
```

Comment:
What kind of benchmarking have you done to validate that this is indeed faster? I ran a test here with a small HTML document and it looks like the old implementation was actually faster on average:

```python
>>> import prairielearn
>>> import timeit
>>> import lxml.html
>>> element = lxml.html.fragment_fromstring('<div>test<p>hello</p>test2</div>')
>>> timeit.timeit(lambda: prairielearn.inner_html_new(element), number = 10000)
0.04604645899962634
>>> timeit.timeit(lambda: prairielearn.inner_html_new(element), number = 10000)
0.04699804197298363
>>> timeit.timeit(lambda: prairielearn.inner_html_new(element), number = 10000)
0.047380250005517155
>>> timeit.timeit(lambda: prairielearn.inner_html(element), number = 10000)
0.02854979201219976
>>> timeit.timeit(lambda: prairielearn.inner_html(element), number = 10000)
0.03979900002013892
>>> timeit.timeit(lambda: prairielearn.inner_html(element), number = 10000)
0.03487775003304705
```

Reply:
Ran this with a large test case:

![image](https://user-images.githubusercontent.com/1345068/211740054-05e501b4-44a8-4094-bef0-dcd7840daae6.png)

And a smaller one:
![image](https://user-images.githubusercontent.com/1345068/211740749-ebd83b1f-ed32-40f4-ac6d-53818245c830.png)


The perf impact is definitely on the subtle side (😅 ), this was more pronounced on my other machine (the one that ran these benchmarks is a very fast workstation). Looking at this again, the main benefit of this PR is really the uplifted test cases for this function and adding in the missing type hints. I think this can probably wait until some of the other element / utility PRs are in. That way there will be a few different files that use the new types before adding this one on top.